src/backend/renderer/element/surface.rs

```diff
@@ -336,31 +346,54 @@ impl<R: Renderer + ImportAll> WaylandSurfaceRenderElement<R> {
         kind: Kind,
     ) -> Result<Self, <R as Renderer>::Error>
     where
-        <R as Renderer>::TextureId: 'static,
+        <R as Renderer>::TextureId: Clone + 'static,
     {
         let id = Id::from_wayland_resource(surface);
         crate::backend::renderer::utils::import_surface(renderer, states)?;
 
+        let data_ref = states
+            .data_map
+            .get::<RendererSurfaceStateUserData>()
+            .map(|data| data.borrow());
+        let mut fallback_data = None;
+
+        let data = data_ref.as_deref().unwrap_or_else(|| {
+            fallback_data = Some(RendererSurfaceState::default());
+            fallback_data.as_ref().unwrap()
+        });
+
         Ok(Self {
             id,
             location,
             alpha,
-            surface: surface.clone(),
             renderer_type: PhantomData,
             kind,
+            view: data.view(),
+            current_commit: data.current_commit(),
+            buffer: data.buffer().cloned(),
+            buffer_scale: data.buffer_scale,
+            buffer_transform: data.buffer_transform,
+            buffer_dimensions: data.buffer_dimensions,
+            damage: data.damage.clone(),
+            opaque_regions: data.opaque_regions.clone(),
```

Comment:
This is more a note to myself, we should probably also provide a COW structure and maybe apply the same optimizations like we did with `DamageSet` for opaque regions.

Reply:
moved to https://github.com/Smithay/smithay/issues/1396